### ISSUE TITLE
Improve the way to use this tool

### ISSUE THREAD

**Issue by jdneo** [jumattos/lsif-util#10]
This is what I most wanted. Let me explain more about this...

I mean, we can have a single main entrance for this tool. Then, for a user, if he installed this module into his global NPM module, he can simply call, for example:
```bash
$> lsif validate /a/b/c/lsif.json
```
to do the validation.

meanwhile, the related API can also be exposed, like we export the function:

```typescript
export async function validate(lsifPath: string): Promise<boolean>
export async function validate(lsifJsonObj: {}): Promise<boolean>
```
Then for the users, they can simply integrate this tool into his own build-tool for testing purpose. For example: `gulp`

@arjun27 @jumattos What do you think?

**Comment by jumattos** [comment]
This is an awesome idea! When I started writing this tool, I just wanted to be able to quickly verify my own LSIF output. Arjun pointed out that it could be useful for other LSIF developers, so I decided to share. I did so in the somewhat rough state that I had it.

Arjun and I agree that your suggestions would polish the tool and increase its impact. We have been discussing offline about the future of this repo, publishing this as a package or even as part of the protocol package (afterall, they are intrinsically related).

Thank you very much for the time taken on this feedback. We will be working on it.

**Comment by jdneo** [comment]
@jumattos Awesome! Looking forward to it!

I'm willing to contribute... 😄

**Comment by jumattos** [comment]
Feel free to contribute! Your feedback so far has been really helpful!

Recently the LSIF protocol I use from `lsif-typescript` has been published to a separate, public package. Right now my backlog looks like this:

1. Validate the whole input regardless of errors (issue #11)
2. Using the new protocol package (so that users won't have to clone `lsif-typescript`)
3. Validate that edges exist only between defined vertices (missing feature)

In other words, I am prioritizing polishing the tool before making it into an API/adding a better CLI. I've also been refactoring some of my code to make it more readable and easier to contribute to.

I am the only dev working on this tool right now, so contributions are most welcome!

**Comment by jumattos** [comment]
Hey @jdneo! `lsif-typescript` has been through some structural changes and is now known as [lsif-node](https://github.com/Microsoft/lsif-node). The tools in this repo will likely migrate to that one.

I have been thinking about the user experience and I believe your insight (since you were the one who brought it up) could be very helpful. I am also pinging @arjun27 for his feedback.

This is the flow I have been thinking of for the CLI:

> `npm install -g lsif-tools`
> `lsif-validate .\output.json`
> `lsif-search .\output.json --type edge`
> `lsif-graph .\output.json --root 10`

Top of my head right now:

- Making the arguments more friendly (no need to pass -p --inputPath for example)
- Improve the validation (edges between defined vertices)
- More options for graph output (other formats besides dot)

Let me know what you think!

**Comment by jdneo** [comment]
Hi @jumattos,

The workflow looks good to me. I have one suggestion, please correct me if I'm wrong.

It seems that the current implementation will output the final result in the console according to https://github.com/jumattos/lsif-util/blob/master/src/validate.ts#L238-L255.

Since our goal is to integrate the validation tool into the test pipeline. I think the tool could throw an error if there is any invalidated vertices or edges. Otherwise, the test pipeline needs to parse the output to know if the index file is correct or not.

**Comment by jumattos** [comment]
You are absolutely right. However, the way I wrote this code, we prepare the output before printing. It would be easy to just skip the printing and throw the errors at [line 94](https://github.com/jumattos/lsif-util/blob/7ab5db3fc6fe7424ad14a4efd07b5bffbecaa079/src/validate.ts#L94).

Right now the tool is returning `true`/`false`. It would be more appropriate to return an error code. We can have a flag to suppress output if needed.

**Comment by arjunattam** [comment]
Yes, I agree we need to output an error exit code (instead of true/false).

@jdneo - I really like the idea of using this tool as a part of your CI pipeline. Please help us prioritize the right experience for this use-case. My understanding of the scenario is:

1. The LSIF exporter tool outputs one or more files
2. You acquire the `lsif-util` tool from NPM
3. You run `lsif-util` validation on the generated files and see if they pass

Is this how you are planning to use it? What is the additional value of getting the validation exposed as JS APIs?

**Comment by jdneo** [comment]
@arjun27 Yes, that's exactly what I'm thinking about. Thanks!

**Comment by jumattos** [comment]
Hey @jdneo! Would you mind telling me which environment are you using for development?

I had some problems trying to output the exit code to PowerShell.

**Comment by jumattos** [comment]
Hello @jdneo! As we discussed offline, the updates that I have been meaning to make are finally out! This means:

- Better interface and easier commands
- Exit code that reflects whether or not validation failed

For the next month, I want to move the code to `lsif-node` and promptly publish it as an npm package. Stay tuned!

Thank you for your continuous feedback and support. Don't hesitate to reach out if something doesn't work as expected with the new version of the tool!

**Comment by jdneo** [comment]
Awesome! Can wait to integrate it into the test pipeline.

Thank you @jumattos

**Comment by jumattos** [comment]
@jdneo You can start using it as soon as you'd like! 😊 Here is the [npm package](https://www.npmjs.com/package/lsif-util).

Since the tool is moving soon to [lsif-node](https://github.com/microsoft/lsif-node), I will close this issue. Thanks again!
